src/solvers/refinement/string_builtin_function.cpp

```diff
@@ -0,0 +1,209 @@
+/// Module: String solver
+/// Author: Diffblue Ltd.
+
+#include "string_builtin_function.h"
+
+#include <algorithm>
+#include "string_constraint_generator.h"
+
+/// Get the valuation of the string, given a valuation
+static optionalt<std::vector<mp_integer>> eval_string(
+  const array_string_exprt &a,
+  const std::function<exprt(const exprt &)> &get_value);
+
+/// Make a string from a constant array
+static array_string_exprt make_string(
+  const std::vector<mp_integer> &array,
+  const array_typet &array_type);
+
+string_transformation_builtin_functiont::
+  string_transformation_builtin_functiont(
+    const std::vector<exprt> &fun_args,
+    array_poolt &array_pool)
+{
+  PRECONDITION(fun_args.size() > 2);
+  const auto arg1 = expr_checked_cast<struct_exprt>(fun_args[2]);
```

Comment:
Could you document the magic indices and offsets please? A brief layout of the vectors would help.

Reply:
@peterschrammel I now document the constructor in the header, hopefully that should be clearer.